analysis/src/SharedTypes.ml

```diff
@@ -425,6 +428,80 @@ module Completable = struct
     | CPObj of contextPath * string
     | CPPipe of contextPath * string
 
+  (* How to move through a nested type context, like from a root record to the type of one of its fields, and beyond. *)
+  (* TODO: Can extend to tuples, objects, etc *)
+  type patternPathItem =
+    | RField of {fieldName: string; alreadySeenFields: string list}
+
+  let str s = if s = "" then "\"\"" else s
+  let list l = "[" ^ (l |> List.map str |> String.concat ", ") ^ "]"
+  let ident l = l |> List.map str |> String.concat "."
+
+  let recordFieldContextPathToString pathItems =
+    pathItems
+    |> List.map (fun item ->
+           match item with
+           | RField {fieldName; alreadySeenFields} ->
+             fieldName ^ ":" ^ list alreadySeenFields)
+    |> list
+
+  let completionContextToString = function
+    | Value -> "Value"
+    | Type -> "Type"
+    | Module -> "Module"
+    | Field -> "Field"
+
+  let rec contextPathToString = function
+    | CPString -> "string"
+    | CPApply (cp, labels) ->
+      contextPathToString cp ^ "("
+      ^ (labels
+        |> List.map (function
+             | Asttypes.Nolabel -> "Nolabel"
+             | Labelled s -> "~" ^ s
+             | Optional s -> "?" ^ s)
+        |> String.concat ", ")
+      ^ ")"
+    | CPArray -> "array"
+    | CPId (sl, completionContext) ->
+      completionContextToString completionContext ^ list sl
+    | CPField (cp, s) -> contextPathToString cp ^ "." ^ str s
+    | CPObj (cp, s) -> contextPathToString cp ^ "[\"" ^ s ^ "\"]"
+    | CPPipe (cp, s) -> contextPathToString cp ^ "->" ^ s
+
+  (* Temporary while I figure things. Pretty sure this should be integrated into contextPath when this is all finished *)
+  type howToRetrieveSourceType =
```

Comment:
Not too surprising that there's a list of possible patters to be filled.
This is not very different from how now one extract a `contextPath` as "instructions" to be executed by the back-end.